### PR TITLE
[CALCITE-2672] Qualifying the common column should not be allowed in Oracle dialect

### PR DESCRIPTION
**Pull request by chunweilei**
JIRA: https://issues.apache.org/jira/browse/CALCITE-2672

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rex/RexBuilder.java [hunk 1]
```diff
@@ -932,13 +925,6 @@ protected RexLiteral makeLiteral(
...
       break;
     case TIME:
-      assert o instanceof TimeString;
-      p = type.getPrecision();
-      if (p == RelDataType.PRECISION_NOT_SPECIFIED) {
-        p = 0;
-      }
-      o = ((TimeString) o).round(p);
-      break;
     case TIME_WITH_LOCAL_TIME_ZONE:
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rex/RexBuilder.java]: These codes are duplicate with TIME_WITH_LOCAL_TIME_ZONE. So remove it.

File: core/src/main/java/org/apache/calcite/rex/RexBuilder.java [hunk 2]
```diff
@@ -669,16 +669,9 @@ private RexNode makeCastIntervalToExact(RelDataType toType, RexNode exp) {
...
+    final BigDecimal divider = endUnit.multiplier;
     RexNode value = multiplyDivide(decodeIntervalOrDecimal(exp),
         multiplier, divider);
-    if (scale > 0) {
-      RelDataType decimalType =
-          typeFactory.createSqlType(SqlTypeName.DECIMAL,
-              scale + exp.getType().getPrecision(), scale);
-      value = encodeIntervalOrDecimal(value, decimalType, false);
-    }
     return ensureType(toType, value, false);
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/rex/RexBuilder.java]: Dead codes.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java [hunk 3]
```diff
@@ -408,4 +408,14 @@
    * false otherwise.
    */
   boolean allowPluralTimeUnits();
+
+  /**
+   * Whether to allow qualifying common column.
+   *
+   * <p>Among the built-in conformance levels, false in
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java]: What do you mean by "common column" ? Can you clarify ?

File: core/src/main/java/org/apache/calcite/materialize/LatticeSuggester.java [hunk 4]
```diff
@@ -334,7 +334,7 @@ private Lattice findMatch(final Lattice lattice, MutableNode mutableNode) {
       }
 
       if (bestMatch != null) {
-        // Fix up the best batch
+        // Fix up the best match
         for (Path path
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/materialize/LatticeSuggester.java]: There is already a fix there, remove this change or keep credit for the contributor.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 5]
```diff
@@ -5623,6 +5623,33 @@ private void checkNegWindow(String s, String msg) {
...
+    // cannot qualify common column in Oracle.
+    sql("select ^emp.deptno^ from emp join dept using (deptno)")
+        .withConformance(SqlConformanceEnum.ORACLE_10)
+        .fails("Cannot qualify common column 'EMP.DEPTNO'")
+        .withConformance(SqlConformanceEnum.ORACLE_12)
+        .fails("Cannot qualify common column 'EMP.DEPTNO'");
+
+    sql("select ^emp.deptno^ from emp natural join dept")
+        .withConformance(SqlConformanceEnum.ORACLE_10)
+        .fails("Cannot qualify common column 'EMP.DEPTNO'");
```
Comment (yanlin-Lynn) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: duplicate with the case below

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java [hunk 6]
```diff
@@ -5623,6 +5623,33 @@ private void checkNegWindow(String s, String msg) {
...
+
+    sql("select count(^emp.deptno^) from emp join dept using (deptno)")
+        .withConformance(SqlConformanceEnum.ORACLE_10)
+        .fails("Cannot qualify common column 'EMP.DEPTNO'")
+        .withConformance(SqlConformanceEnum.ORACLE_12)
+        .fails("Cannot qualify common column 'EMP.DEPTNO'");
+
+    sql("select emp.deptno from emp join dept using (deptno)")
+        .withConformance(SqlConformanceEnum.DEFAULT)
+        .ok();
```
Comment (yanlin-Lynn) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: can you add case for non common column?

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java [hunk 7]
```diff
@@ -527,24 +527,48 @@ private static SqlNode expandExprFromJoin(SqlJoin join, SqlIdentifier identifier
...
     return expandExprFromJoin((SqlJoin) from, identifier, scope);
   }
 
+  private static void validateQualifiedCommonColumn(SqlJoin join,
+      SqlIdentifier identifier, SelectScope scope, SqlValidator validator) {
+    if (!(join.isNatural() || join.getConditionType() == JoinConditionType.USING)) {
+      return;
+    }
+    for (ScopeChild child : scope.children) {
+      if (child.name.equals(identifier.getComponent(0).toString())) {
```
Comment (yanlin-Lynn) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: if the column is not a common column ,like `select emp.ename from emp join dept using(deptno)`, it should not fail.

File: core/src/main/java/org/apache/calcite/materialize/LatticeSuggester.java (hunk 4 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/materialize/LatticeSuggester.java]: Will do.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java (hunk 5 above)
Comment (chunweilei) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: Oops, will remove it.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java (hunk 7 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: You're right. But it is surprising that we do not have such test case. I'll add one.

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java (hunk 6 above)
Comment (chunweilei) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java]: Will do.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java (hunk 3 above)
Comment (chunweilei) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlConformance.java]: Will do.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java [hunk 8]
```diff
@@ -314,4 +314,18 @@ public boolean allowExtendedTrim() {
...
   }
 
+  @Override public boolean allowQualifyingCommonColumn() {
+    switch (this) {
+    case ORACLE_10:
+    case ORACLE_12:
+    case STRICT_92:
+    case STRICT_99:
+    case STRICT_2003:
+      return false;
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlConformanceEnum.java]: SQL standard also disallows qualifying the common columns[1].

[1] https://dev.mysql.com/doc/refman/8.0/en/join.html

**Pull request merged**